Utxow Predicate failure tests

teodanciu opened the pull request:
# Description

Tests for UTXOW Predicate failures, part of https://github.com/IntersectMBO/cardano-ledger/issues/4185

 InvalidWitnessesUTXOW
 MissingVKeyWitnessesUTXOW
 MissingScriptWitnessesUTXOW
 ScriptWitnessNotValidatingUTXOW
 MissingTxBodyMetadataHash
 MissingTxMetadata
 ConflictingMetadataHash
 InvalidMetadata
 ExtraneousScriptWitnessesUTXOW
 MissingRedeemers
 MissingRequiredDatums
 NotAllowedSupplementalDatums
 PPViewHashesDontMatch
 UnspendableUTxONoDatumHash
 ExtraRedeemers
 MalformedScriptWitnesses
 MalformedReferenceScripts

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

teodanciu committed:
Add a malformed script to Imp `scriptTestContexts`

to enable testing invalid scripts

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . auxDataHashTxBodyL .~ SJust auxDataHash
+            & auxDataTxL .~ SJust auxData
+    submitFailingTx tx [injectFailure InvalidMetadata]
```
The only thing that is preventing you from having this test being era agnostic is lack of lenses, like we do have for other type families.

I think it is time we add them to the interface: #4267 

Let's rebase this PR onto #4267 and consolidate this test and the other from Alonzo into one that works for all eras starting with Allegra. All you need from the aforementioned PR is `mkBasicTxAuxData` and `metadataTxAuxDataL`

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+          [ injectFailure $
+              PPViewHashesDontMatch (SJust wrongIntegrityHash) (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)
+          ]
+    void $ impAnn "Missing" $ do
```
Unnecessary. `submitFailingTx` returns unit.
```suggestion
    impAnn "Missing" $ do
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+    let scriptHash = withSLanguage lang (hashPlutusScript . guessTheNumber3)
+    txIn <- produceScript scriptHash
+    tx <- fixupTx $ mkBasicTx mkBasicTxBody & bodyTxL . inputsTxBodyL .~ [txIn]
+    void $ impAnn "Mismatched " $ do
```
Same as below. Rednundant.
```suggestion
    impAnn "Mismatched " $ do
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+        & nesEsL . esLStateL . lsUTxOStateL . utxosUtxoL
+          %~ ( \(UTxO utxo) ->
+                UTxO (Map.adjust resetDatum txIn utxo)
+             )
```
We should try our best to avoid modifying the ledger state. Why don't you instead use `withPostFixup` on the first tx (that is done by `produceScript`) to remove the data hash from the TxOut?

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+      tx
+      [ injectFailure $
+          ExtraRedeemers [prp]
+      ]
```
The line isn't that long when we squash it. One line is better than 5 if it is still a readable line :slightly_smiling_face: 
```suggestion
    submitFailingTx tx [injectFailure $ ExtraRedeemers [prp]]
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+    submitFailingTx tx [injectFailure InvalidMetadata]
+  where
+    fixupResetAddrWits = fixupPPHash >=> resetAddrWits
+    resetAddrWits tx = updateAddrTxWits $ tx & witsTxL . addrTxWitsL .~ []
```
Let's put these at the very top as let bindings, this will be a much clear indication of their scope and where these helpers are coming from in their use sites:

```haskell
describe "UTXOW" $ do
  let fixupResetAddrWits = fixupPPHash >=> resetAddrWits
      resetAddrWits tx = updateAddrTxWits $ tx & witsTxL . addrTxWitsL .~ []
  it "MissingRedeemers" $ do
    ...
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+    submitFailingTx tx [injectFailure InvalidMetadata]
+
+genInvalidMetadatum :: ImpTestM era (Map.Map Word64 Metadatum)
+genInvalidMetadatum = do
```
`Matadatum` is a single item. The field that contains a Map is called `metadata` :slightly_smiling_face: 
```suggestion
genInvalidMetadata :: ImpTestM era (Map.Map Word64 Metadatum)
genInvalidMetadata = do
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+          keyPair <- impRootKh >>= lookupKeyPair
+          (addRootTxIn >=> fixupFees) tx <&> withWitVKey keyPair
+    void $
+      withFixup fixup $
```
Redundant `void` and also why not use `withPostFixup` here as well?
```suggestion
    withPostFixup (\tx -> tx & inputsTxBodyL . bodyTxL <>~ [txIn])
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp.hs:
```diff
@@ -1,23 +1,35 @@
-  ( ShelleyEraImp era
-  , AlonzoEraTx era
+  ( AlonzoEraImp era
+  , TxAuxData era ~ AlonzoTxAuxData era
```
#4267 will let you get rid of this constraint
```suggestion
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+  , InjectRuleFailure "LEDGER" ShelleyUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
+  , Arbitrary (Data era)
```
Import the orphan instances: `Test.Cardano.Ledger.Alonzo.Arbitrary` and this constraint will be satisfied
```suggestion
```

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp.hs:
```diff
@@ -1,23 +1,35 @@
+  , InjectRuleFailure "LEDGER" ShelleyUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxowPredFailure era
+  , InjectRuleFailure "LEDGER" AlonzoUtxosPredFailure era
+  , Arbitrary (Data era)
```
Redundant
```suggestion
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+specAllegra ::
+  forall era.
+  ( ShelleyEraImp era
+  , TxAuxData era ~ AllegraTxAuxData era
```
See my comment below.
```suggestion
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+  let genM =
+        oneof
+          [ B . BS.pack <$> vectorOf size arbitrary
+          , S . T.pack <$> vectorOf size arbitrary
```
This will be a bit more strict, since the size for Text is counted in number of bytes, not number of characters.
```suggestion
          , S . T.pack <$> genUtf8StringOfSize size
```

lehins commented on eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,54 @@
+          MalformedScriptWitnesses [scriptHash]
+      ]
+
+  it "MalformedReferenceScripts" $ do
```
Let's get rid of the old test that this ones replaces:

* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs#L1397-L1404
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs#L877-L902

lehins commented on eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,54 @@
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "UTXOW" $ do
+  it "MalformedScriptWitnesses" $ do
```
Let's get rid of this one:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs#L1405-L1418
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/BabbageFeatures.hs#L1010-L1043

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+              PPViewHashesDontMatch SNothing (tx ^. bodyTxL . scriptIntegrityHashTxBodyL)
+          ]
+
+  it "UnspendableUTxONoDatumHash" $ do
```
Let's get rid of:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L324-L331
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L805-L833

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+          NotAllowedSupplementalDatums [extraDatumHash] []
+      ]
+
+  it "PPViewHashesDontMatch" $ do
```
Let's get rid of:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L185-L207
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L448-L462

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+            MissingRequiredDatums [missingDatum] []
+        ]
+
+  it "NotAllowedSupplementalDatums" $ do
```
Let's get rid of this one:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L332-L344
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L835-L862

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+            CollectErrors [NoRedeemer missingRedeemer]
+        ]
+
+  it "MissingRequiredDatums" $ do
```
Let's get rid of this one:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L245-L255
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L624-L652

lehins commented on eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,168 @@
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "UTXOW" $ do
+  it "MissingRedeemers" $ do
```
Let's get rid of this one:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L175-L184
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/AlonzoInvalidTxUTXOW.hs#L415-L446

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxowSpec.hs:
```diff
@@ -49,3 +56,52 @@ spec = describe "UTXOW" $ do
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . outputsTxBodyL .~ [mkBasicTxOut bobAddr (inject (Coin 10))]
+    void $
+      withFixup (addRootTxIn >=> fixupFees) $
```
Redundant `void` and relying on root TxIn is a bit convoluted. Why not create a new txIn and use a postfixup? I think something like this should work:
```suggestion
    aliceKh <- freshKeyHash
    txIn <- sendCoinTo (Addr Testnet (KeyHashObj aliceKh) StakeRefNull) 99
    withPostFixup (\tx -> tx & inputsTxBodyL . bodyTxL <>~ [txIn]) $
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+      tx & witsTxL . addrTxWitsL .~ [witVKey keyPair tx]
+    witVKey keyPair tx =
+      let bodyHash = hashAnnotated $ tx ^. bodyTxL
+       in mkWitnessVKey bodyHash keyPair
```
These are only used in one test, no need to make them available in other tests. Also, I think it is better to defined such functions at the top with a `let` for consistency and readibility.
Moreover, I think these two can be deleted if my suggestion using `withPostFixup` will work

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+    let scriptHash = hashScript @era script
+    let tx = mkBasicTx mkBasicTxBody & witsTxL . scriptTxWitsL .~ [(scriptHash, script)]
+    withFixup shelleyFixupTx $
+      submitFailingTx
```
Why customization of the fixup. Shouldn't it work with a usual fixup function, since presupplied witnesses will not be removed?
```suggestion
    submitFailingTx
```

lehins commented on eras/allegra/impl/testlib/Test/Cardano/Ledger/Allegra/Imp/UtxowSpec.hs:
```diff
@@ -0,0 +1,100 @@
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "UTXOW" $ do
+  it "MissingScriptWitnessesUTXOW" $ do
```
Both `MissingScriptWitnessesUTXOW` and `ExtraneousScriptWitnessesUTXOW` are applicable to `Shelley` as well, thus they can be moved into the Shelley imp spec instead.

Also once you do that we can remove this test:
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs#L169-L183
* https://github.com/IntersectMBO/cardano-ledger/blob/2c9ba10d085331fa99840a63b40b8e8b716fba5d/eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs#L94

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -729,6 +730,14 @@ lookupImpRootTxOut = do
+  pure $ case rootTxOut ^. addrTxOutL @era of
+    Addr _ (KeyHashObj kh) _ -> kh
+    _ -> error "Missing expected root key hash"
+
```
I think this approach is little too hacky. We should not use the root tx for more than one purpose, especially if it can be avoided
```suggestion
```

lehins reviewed (changes requested):
This is really great work. There are a bunch of comments, but I think they all can improve the PR even more than it already is.

teodanciu committed:
Add produceScript to shelley ImpTest and use it

teodanciu committed:
Add tests for UTXOW Predicate failures for allegra+ eras

teodanciu committed:
Add tests for UTXOW Predicate failures for alonzo+ eras

teodanciu committed:
Add ImpTestEnv flag to expect failures in pred failures roundtrip test

in order to support testing failures for which we know that the
roundtripping is not working (and needs to be kept like this for
backwards compatibility)

teodanciu committed:
Add tests for UTXOW Predicate failures for shelley+ era

teodanciu committed:
Add tests for UTXOW Predicate failures for babbage+ eras

teodanciu committed:
Remove tests made redundant by the newly introduced UTXOW Imp tests

teodanciu commented:
I have rewritten most of these tests to make more use of postFixup.  In the initial version, I was trying to produce the exact list of failures, and for that, the fixing up was more complicated and less comprehensible. 
I  have now switched to checking that the predicate failure under test is included in all failures, which allowed me an easier fixup. 
Also, moved things around in order to run more tests, thanks to the new  EraTxAuxData and EraScript instances. 

If you wanted to have another look @lehins  when you have time. Also, @neilmayhew   - hope this isn't too much interfering with the work on your other (related) PR.

neilmayhew commented on eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs:
```diff
@@ -332,7 +333,7 @@ instance Era era => DecCBORGroup (AlonzoPlutusPurpose AsIx era) where
     decodeWord8 >>= \case
       0 -> AlonzoSpending . AsIx <$> decCBOR
-      1 -> AlonzoMinting . AsIx <$> decCBOR
+      1 -> trace "HERE!!!!" (AlonzoMinting . AsIx <$> decCBOR)
```
It looks like this got left in by mistake :smile:

teodanciu replied:
haha, oops!  Thank you

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -974,6 +987,20 @@ submitFailingTx ::
+    >>= expectLeftExpr
+    >>= ( `shouldSatisfy`
+            (\fs -> all (`elem` fs) expectedFailuresSubset)
+        )
```
I am not sure I like this approach of testing only presence of some of the failures. 
By checking the exact sequence of predicate failures we are testing the exact behavior of the rule. If a rule has a slightly different outcome from one era to another then we should have separate tests for such rule.

I understand that it is easier to solve it this way, but I believe this is a dangerous slope. Moreover, we are translating some of the tests that checked exact behavior, thus in this PR we are making our test suite weaker, rather than better. 

More importantly, we have all of the tooling necessary to deal with this sort of discrepancies between eras in a more accurate way. See my suggestion on the `ExtraneousScriptWitnessesUTXOW` test for an example.

So, could you please remove this `submitFailingTxWithSubset` and switch back to `submitFailingTx` instead

teodanciu replied:
I agree.  Removed `submitFailingTxWithSubset`  and actually it wasn't too bad to do without it, thank you.

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/Imp/UtxowSpec.hs:
```diff
@@ -49,3 +61,69 @@ spec = describe "UTXOW" $ do
+      tx
+      [ injectFailure $
+          ExtraneousScriptWitnessesUTXOW [scriptHash]
+      ]
```
This is a more accurate way of testing predicate failures that also nicely depict changes between eras:
```suggestion
        -- We dropped validating scripts when they are not needed starting with Babbage
        expectedFailures
          | eraProtVerLow @era >= natVersion @6 =
              [ injectFailure $ ExtraneousScriptWitnessesUTXOW [scriptHash]
              ]
          | otherwise =
              [ injectFailure $ ScriptWitnessNotValidatingUTXOW [scriptHash]
              , injectFailure $ ExtraneousScriptWitnessesUTXOW [scriptHash]
              ]
    submitFailingTx tx expectedFailures
```

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -938,10 +947,14 @@ trySubmitTx tx = do
+      when roundTripCheck $
+        liftIO $
+          forM_ predFailures $
+            roundTripEraExpectation @era
```
It is better to ensure that there is a roundtrip failure, otherwise we might be either mistakenly disabling the check or we have already fixed the serialization, but forgot to adjust the test.
```suggestion
      liftIO $
        if roundTripCheck
          then
            forM_ predFailures $ roundTripEraExpectation @era
          else
            roundTripCborRangeFailureExpectation
              (eraProtVerLow @era)
              (eraProtVerHigh @era)
              predFailures
```
Note that we need to roundtrip the whole list, cause we can't specify which one of the failures is faulty

teodanciu replied:
Nice, thank you. In this way, I was forced to not skip the tests for Conway, where the serialization was fixed.

lehins reviewed (changes requested):
Ok, I do have a problem with `submitFailingTxWithSubset`. There is a proposed solution on how to solve this in the review. 
Also the PredicteFailrues that do not roundtrip as expected can be tested for that as well.

Other than that this PR is looking much better than before :wink: thanks to all of the work you've done in other PRs.

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -938,10 +948,22 @@ trySubmitTx tx = do
+      when roundTripCheck $
+        liftIO $
+          forM_ predFailures $
+            roundTripEraExpectation @era
```
Duplicate check
```suggestion
```

lehins commented on eras/shelley/impl/testlib/Test/Cardano/Ledger/Shelley/ImpTest.hs:
```diff
@@ -938,10 +948,22 @@ trySubmitTx tx = do
+              (eraProtVerLow @era)
+              (eraProtVerHigh @era)
+              predFailures
       -- Verify that produced predicate failures are ready for the node-to-client protocol
```
Let's move the comment above
```suggestion
      -- Verify that produced predicate failures are ready for the node-to-client protocol
      if roundTripCheck
        then liftIO $ forM_ predFailures $ roundTripEraExpectation @era
        else
          liftIO $
            roundTripCborRangeFailureExpectation
              (eraProtVerLow @era)
              (eraProtVerHigh @era)
              predFailures
```

lehins reviewed (approved):
Beautiful! Thank you!
